doc/developer/design/20240919_dataflow_expiration.md

```diff
@@ -56,6 +102,34 @@ the dataflow expiration are dropped. To ensure correctness, panic checks are
 added to these dataflows that ensure that the frontier does not exceed the
 dataflow expiration before the replica is restarted.
 
+Environment and replica only have partial information on the per-dataflow
+expiration time. The environment knows the dependency tree and properties per
+object. Only the replica knows the local expiration time. Thus, a dataflow
+description needs to encode enough breadcrumbs for the replica to make a
+correct decision on if and when to expire a dataflow.
+
+**An object is _definite up to_ a frontier as a function of an expiration
+time:**
+* Sources are definite up to the expiration time.
+* Load generators are definit up to the minimum frontier.
+* Indexes are definite up to the meet of their inputs.
+* Constant collections are definite up to the empty frontier.
+* Tables are definite up to the expiration time.
+* Materialized views are definite up to the meet of their inputs, rounded up to
+  the refresh schedule.
+* Subscribes are definite up to the meet of their up-to and their inputs.
+
+### Distinguishing expiration from dataflow shutdown
+
+When a dataflow shuts down, it frontier advances to the empty frontier. The
+persist source assumes that once the input frontiers reach the `until`, it is
+free to shut down the dataflow. This is correct if the until is specified by
+the user, but incorrect if the until is set to expire future updates.
+
+TODO:
+* Checking frontiers at the input vs. output.
+* Do we need a token to distinguish shutdown variants? Ideally, we know from the structure of the dataflow and objects it depends on what advancing to the empty frontier means.
```

Comment:
Unfortunately, I don't believe this will work. We still need to distinguish a dataflow advancing to the empty frontier because of reaching the until, or because the inputs shut down. I'm thinking of a better alternative that will avoid the issue in the first place.